XiEditor/EditView.swift

```diff
@@ -146,15 +146,18 @@ class EditView: NSView, NSTextInputClient {
         let first = max(0, Int((floor(dirtyRect.origin.y - topPad) / dataSource.textMetrics.linespace)))
         let last = Int(ceil((dirtyRect.origin.y + dirtyRect.size.height - topPad) / dataSource.textMetrics.linespace))
 
-        let missing = dataSource.lines.computeMissing(first, last)
-        for (f, l) in missing {
-            Swift.print("requesting missing: \(f)..\(l)")
-            dataSource.document.sendRpcAsync("request_lines", params: [f, l])
+        let lines = dataSource.lines.blockingGet(lines: first..<last)
+
+        let missing = lines.enumerated().filter( { $0.element == nil } )
+            .map( { $0.offset + first } )
+        if !missing.isEmpty {
+            print("draw missing lines: \(missing)")
         }
 
         // first pass, for drawing background selections and search highlights
-        for lineIx in first..<last {
-            guard let line = getLine(lineIx), line.containsReservedStyle == true else { continue }
+        for lineIx in first...last {
```

Comment:
I became very confused about how we describe line ranges while doing this work. (The `LineRange` type alias was partly a response to this). There's also confusion because we calculate the line range for a given rect differently in different places (drawrect vs willScroll).

Anyway: good spot, the oob was always getting caught in the guard statement.